GoogleForCreators/web-stories-wp
# Pre-Publish Checklist: select offending elements from tab

## Summary
- This is step 1 of a 2-step enhancement to select and highlight the offending elements in the editor when clicking on the checklist. (details in ticket [here](https://app.zenhub.com/workspaces/web-stories-5e94d3aced449034e1e9b226/issues/google/web-stories-wp/5629))
- I wrote a doc to be reviewed by UX [here](https://docs.google.com/document/d/1soSgYqJoIRwiyWZxs2FBVTuw2-EQf4KJQQG3HBnUhM4/edit) so that the behavior could be reviewed outside of the code review. The designs for the feature are still TBD.


## Relevant Technical Choices
- This first implementation I wanted to keep simple, so the logic lives in the checklist tab for now, where we can use hooks to send the user to the correct page and highlight the elements. We were already providing the element ID and page ID from the checks so it was just a matter of providing those details.
- I made changes to the return object for a couple of the checks so that the checklist could highlight multiple elements when it's not so clear what elements are offending the checklist--"page has too much text", for example.
- This step does not include opening the correct panels or highlighting specific parts of the editor. It just selects the offending elements. I think this looks confusing if you're already selecting the offending element. And also, not everything is clickable yet... Nor is it obvious that a user could click on the rows in the checklist tab. I think there are multiple ways to improve this incrementally, probably by adding a new "highlighted" state which can put notices or styles next to/on top of elements as well as panels and inputs in the editor. Working in the background on improving the way we run the checklist so that the side-effects are cheaper and we could potentially have these highlights appear live (not recommended at the moment). I have an [extremely rough draft here](https://docs.google.com/document/d/1-9mnhzx28dFtlPI3gFFL3YDHX16nyy_F5hGL7Ajg-RI/edit?usp=sharing) where I've started collecting my thoughts about that problem. 


## To-do
- [x] karma test - text (select multiple elements)
- [x] karma test - media (select single element)
- [x] karma test - keyboard controls
- [x] keyboard controls



## User-facing changes
- When hovering on a row in the checklist tab that will highlight elements when clicked, a user will see a pointer cursor instead of their normal cursor.
- Clicking on certain rows in the checklist tab should navigate the user to the page it's concerned about, and select all the elements that the checklist is describing.


## Testing Instructions
1. Create a new story
2. Add an element to the editor that offends the checklist e.g. add a few text boxes with lots of text, add 4-5 element links to a page, a low resolution image, a video, etc.
3. Open the checklist tab-- there should be checks there that are describing the elements that were added in step 2. Clicking on the checklist tab row should open the correct page and select all the offending elements. i.e. if you added 4-5 links to a single page, it should select all the links



---



Fixes #5629
